### PR TITLE
Adding Prometheus and Grafana for load analysis and reimbursements analytics

### PR DESCRIPTION
**Pull request by ArthurSens**
**What is the purpose of this Pull Request?**

Provide a tool that is capable to monitor the performance of all the applications that are part of the serenata solution and show those informations with graphics that we can use to compare the performance of different solutions.

At the same time, with prometheus+grafana, we can provide dashboards that can be used to study the reimbursements data. Prometheus is great for analytics with their [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) while grafana is great for show data with a friendly interface.

**What was done to achieve this purpose?**

I've added prometheus and grafana to the docker-compose, along with some agents that are responsible to provide monitoring metrics from the applications to the prometheus server.

One agent in particular is responsible for quering postgres once a day to provide the reimbursements data to prometheus, so we can make our analytical study on that data.

**How to test if it really works?**

First start all applications
```
docker-compose up -d
```
Your prometheus will be available at http://localhost:9090
Your grafana will be available at http://localhost:3000
You can log in into grafana with the credentials *admin/admin* (It can be changed)

You can see the dashboards at http://localhost:3000/dashboards
You will find 2 folders, `Monitoring` will have all the dashboards responsible to monitor and provide load analysis of serenata's applications and `Business Inteligence` will have the dashboards the we will use to make analytics studies about the reimbursements.
PS: Sometimes the folders and dashboards takes a long time to be provisioned after you start the grafana server. If it's not there, wait 5 to 10 minutes for them to load.

When prometheus and grafana are up, you can monitor the performance of your applications at:
* Linux host -> http://localhost:3000/d/m2nXDoemz/linux?orgId=1
* Postgres -> http://localhost:3000/d/sadfsdfsfd/postgres?orgId=1
* All processes -> http://localhost:3000/d/lOXQMRMZ1/process-monitoring?orgId=1

For load analysis, for example, open the Process monitoring dashboard and then execute:
```
docker run --rm -v /tmp/serenata-data:/tmp/serenata-data serenata/rosie python rosie.py run chamber_of_deputies 
```
You will be able to see how much hardware resources each process is consuming, as show at the image below:

image

If you make changes at rosie's code, you will be able to evaluate if the performance got better or worse.

---
To validate the reimbursements stuff is a lit bit trickier, since it can take 1 whole day to prometheus to gather the data, not because it's slow, just because it gather that information only once a day.

But first you gotta populate the the database.

```
docker-compose run --rm django python manage.py migrate
docker-compose run --rm django python manage.py reimbursements /mnt/data/reimbursements_sample.csv
docker-compose run --rm django python manage.py companies /mnt/data/companies_sample.xz
```

After that you can see if prometheus got that data from postgres if you go to http://localhost:9090/targets and you see the businessInteligence job with the `UP` status. As shown in the image below:
image

If you see the `UNKNOWN` status, it means that prometheus still haven't queried postgres yet. 

Once prometheus have the reimbursement data, you will be able to manipulate that information with PromQL and build dashboards like the one I made, accesible at http://localhost:3000/d/ecOrj5MZz/reimbursements-analytics?orgId=1

**Who can help reviewing it?**

Anyone capable of starting the containers.

**TODO**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -32,6 +32,56 @@ services:
   rosie:
     image: serenata/rosie
 
+  prometheus:
+    image: arthursens/serenata-prometheus
```
Comment (ArthurSens) [inline docker-compose.yml]: arthursens/serenata-prometheus image was built from [this Dockerfile](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-831949f2ad1dcdf050d81cf5a100bc7eR1)

File: docker-compose.yml [hunk 2]
```diff
@@ -32,6 +32,56 @@ services:
...
+    ports: 
+    - 9090:9090
+    links:
+    - node_exporter:node_exporter
+    - process_exporter:process_exporter
+    - postgres_exporter:postgres_exporter
+    - bi_exporter:bi_exporter
+
+  grafana:
+    image: arthursens/serenata-grafana
```
Comment (ArthurSens) [inline docker-compose.yml]: arthursens/serenata-grafana was built from [this Dockerfile](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-933098848ceee5eb4c5202f8be105679R1)

File: docker-compose.yml [hunk 3]
```diff
@@ -32,6 +32,56 @@ services:
...
+    links:
+    - prometheus:prometheus
+    - postgres:postgres
+    ports:
+    - 3000:3000
+
+  node_exporter:
+    image: prom/node-exporter
+
+  process_exporter:
```
Comment (ArthurSens) [inline docker-compose.yml]: This agent is responsible to expose metrics about resources used by every process that runs in your host.
More information about it can be found at [this repository](https://github.com/ncabatoff/process-exporter)

File: docker-compose.yml [hunk 4]
```diff
@@ -32,6 +32,56 @@ services:
...
+    image: ncabatoff/process-exporter
+    privileged: yes
+    volumes: 
+    - /proc:/host/proc
+    - ./prometheus/process-exporter:/config
+    command: --procfs /host/proc -config.path /config/config.yml
+
+  postgres_exporter:
+    image: wrouesnel/postgres_exporter
```
Comment (ArthurSens) [inline docker-compose.yml]: This agent is responsible to execute queries at our postgres and expose the results.
It already have some monitoring queries that are executed by default and more can be added with a YAML file.

I've added some monitoring queries [here](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-5591516ea0bb28550071905892354650R1).

More information about this exporter can be found at [this repository](https://github.com/wrouesnel/postgres_exporter)

File: docker-compose.yml [hunk 5]
```diff
@@ -32,6 +32,56 @@ services:
...
+
+  bi_exporter:
+    image: wrouesnel/postgres_exporter
+    environment:
+      DATA_SOURCE_NAME: postgres://jarbas:mysecretpassword@postgres/jarbas?sslmode=disable
+    links:
+    - postgres:postgres
+    volumes:
+    - ./prometheus/postgres-exporter:/prometheus
+    command: --disable-default-metrics --disable-settings-metrics --extend.query-path /prometheus/businessInteligence.yml
```
Comment (ArthurSens) [inline docker-compose.yml]: This agent uses the same image as [postgres_exporter](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-4e5e90c6228fd48698d074241c2ba760R65), but it has different configurations.

First I've disabled the monitoring queries with the `--disable-default-metrics --disable-settings-metrics` flags
The YAML file with the custom flags is also different, as you can see [here](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-1f8c3941afc2afb592b6ca21368a4961R1)

File: grafana/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,3 @@
+FROM grafana/grafana:6.1.3
+
+ADD ./provisioning /etc/grafana/provisioning
```
Comment (ArthurSens) [inline grafana/Dockerfile]: We will use [Grafana Provisioning](https://grafana.com/docs/administration/provisioning/) to provide our dashboards automatically.

Grafana reads yaml files located at `/etc/grafana/provisioning` to provide dashboards, so we gotta add those files to our grafana image.

File: grafana/provisioning/dashboards/businessInteligence.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+providers:
+- name: 'businessInteligence'
+  orgId: 1
+  folder: 'Business Inteligence'
+  type: file
+  disableDeletion: true
+  updateIntervalSeconds: 300
+  options:
+    path: /etc/grafana/provisioning/dashboards/businessInteligence
+
```
Comment (ArthurSens) [inline grafana/provisioning/dashboards/businessInteligence.yml]: This yaml will create the `Business Inteligence` folder.
All the json files included at the folder `/etc/grafana/provisioning/dashboards/businessInteligence` will be provisioned automatically at the `Business Inteligence` folder.

So if you want to add new dashboards, you just gotta add their corresponding JSONs files at this directory.

File: grafana/provisioning/dashboards/monitoring.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+- name: 'monitoring'
+  orgId: 1
+  folder: 'Monitoring'
+  type: file
+  disableDeletion: true
+  updateIntervalSeconds: 300
+  options:
+    path: /etc/grafana/provisioning/dashboards/monitoring
+
+
```
Comment (ArthurSens) [inline grafana/provisioning/dashboards/monitoring.yml]: This yaml will create the `Monitoring` folder.
All the json files included at the folder `/etc/grafana/provisioning/dashboards/monitoring` will be provisioned automatically at the `Monitoring` folder.

So if you want to add new dashboards, you just gotta add their corresponding JSONs files at this directory.

File: grafana/provisioning/datasources/postgres.yml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+apiVersion: 1
+
+datasources:
+  - name: Postgres
+    type: postgres
+    url: postgres
```
Comment (ArthurSens) [inline grafana/provisioning/datasources/postgres.yml]: The URL to our postgres is just `postgres` thanks to the address we created [here](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-4e5e90c6228fd48698d074241c2ba760R49)

File: grafana/provisioning/datasources/postgres.yml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    user: jarbas
+    secureJsonData:
+      password: "mysecretpassword"
+    jsonData:
+      sslmode: "disable" # disable/require/verify-ca/verify-full
+      maxOpenConns: 0         # Grafana v5.4+
+      maxIdleConns: 2         # Grafana v5.4+
+      connMaxLifetime: 14400  # Grafana v5.4+
+      postgresVersion: 904 # 903=9.3, 904=9.4, 905=9.5, 906=9.6, 1000=10
+      timescaledb: false
```
Comment (ArthurSens) [inline grafana/provisioning/datasources/postgres.yml]: This YAML will provide out Postgres datasource.

Particularly, I haven't used for anything, but you can build dashboards about reimbursements with this datasource.

File: grafana/provisioning/datasources/prometheus.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+datasources:
+- name: Prometheus
+  type: prometheus
+  access: server
+  orgId: 1
+  url: http://localhost:9090
+  version: 1
+  editable: true
+
```
Comment (ArthurSens) [inline grafana/provisioning/datasources/prometheus.yml]: This YAML will provide our Prometheus datasource.

We will use this datasource to build our monitoring and load analysis dashboards, and potencially our Reimbursements analytics dashboards too.

File: prometheus/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,5 @@
+FROM prom/prometheus:v2.10.0
+
+ADD ./prometheus.yml /etc/prometheus/prometheus.yml
+
```
Comment (ArthurSens) [inline prometheus/Dockerfile]: Prometheus needs a configuration file to gather data.

At that file we will tell prometheus where he must scrape data and how often he will scrape it.

This configuration file can be found [here](https://github.com/okfn-brasil/serenata-de-amor/pull/484/files#diff-30171a37a35ae9a4f53c346d781da3efR1)

File: prometheus/postgres-exporter/businessInteligence.yml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    - supplier:
+        usage: "LABEL"
+        description: "Company or person whom provided the receipt"
+    - congressperson_name:
+        usage: "LABEL"
+        description: "Deputie who requested the reimbursement"
+    - total_net_value:
+        usage: "GAUGE"
+        description: "Reimbursement value"
+
```
Comment (ArthurSens) [inline prometheus/postgres-exporter/businessInteligence.yml]: Extra queries can be added here so prometheus gather more analytical information.

File: prometheus/prometheus.yml [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  alertmanagers:
+  - static_configs:
+    - targets: []
+    scheme: http
+    timeout: 10s
+scrape_configs:
+
+- job_name: monitoring
+  honor_timestamps: true
+  scrape_interval: 15s
```
Comment (ArthurSens) [inline prometheus/prometheus.yml]: Monitoring information is collected every 15s

File: prometheus/prometheus.yml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  scrape_timeout: 10s
+  metrics_path: /metrics
+  scheme: http
+  static_configs:
+  - targets:
```
Comment (ArthurSens) [inline prometheus/prometheus.yml]: Address where prometheus scrapes monitoring information

File: prometheus/prometheus.yml [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+- job_name: businessInteligence
+  honor_timestamps: true
+  scrape_interval: 1d
+  scrape_timeout: 30m
+  metrics_path: /metrics
+  scheme: http
+  static_configs:
+  - targets:
+    - bi_exporter:9187
```
Comment (ArthurSens) [inline prometheus/prometheus.yml]: Address where prometheus scrapes analytical information

File: prometheus/prometheus.yml [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  - targets:
+    - localhost:9090
+    - node_exporter:9100
+    - process_exporter:9256
+    - postgres_exporter:9187
+
+
+- job_name: businessInteligence
+  honor_timestamps: true
+  scrape_interval: 1d
```
Comment (ArthurSens) [inline prometheus/prometheus.yml]: Analytical information is collected once a day

**Comment by ArthurSens** [comment]
Prometheus is an [open-source](https://github.com/prometheus/prometheus) monitoring tool and really well accepted by the community.

It uses agents that exposes metrics at a certain port, and then prometheus scrapes those metrics via HTTP requests.

Prometheus has [several agents](https://prometheus.io/docs/instrumenting/exporters/)(they call them 'exporters') build by themselves or by the community and it also have [libraries](https://prometheus.io/docs/instrumenting/clientlibs/) so we can build our own custom exporter, so it can monitor almost everything.

On this PR I've added a [process-exporter](https://github.com/ncabatoff/process-exporter), [postgres-exporter](https://github.com/wrouesnel/postgres_exporter) and [linux-exporter](https://github.com/prometheus/node_exporter), but there is some other agents that can monitor [RabbitMQ](https://github.com/kbudde/rabbitmq_exporter), [Memcached](https://github.com/prometheus/memcached_exporter), [New Relic](https://github.com/mrf/newrelic_exporter) and [Django](https://github.com/korfuri/django-prometheus). I just haven't had the time to build a dashboard to them... if anyone feel confortable to contribute with that, just add the exporter to the docker-compose.yml, build a dashboard and put the JSON file at the grafana monitoring folder.

File: prometheus/postgres-exporter/businessInteligence.yml [hunk 18]
```diff
@@ -0,0 +1,31 @@
+chamber_of_deputies_reimbursement:
+  query: "select distinct id, year, case when month = 1 then 'JANUARY' when month = 2 then 'FEBRUARY' when month = 3 then 'MARCH' when month = 4 then 'APRIL' when month = 5 then 'MAY' when month = 6 then 'JUNE' when month = 7 then 'JULY' when month = 8 then 'AUGUST' when month = 9 then 'SEPTEMBER' when month = 10 then 'OCTOBER' when month = 11 then 'NOVEMBER' when month = 12 then 'DECEMBER' end as month , state, party, subquota_description, supplier, congressperson_name, total_net_value from chamber_of_deputies_reimbursement"
+  metrics:
+    - id:
+        usage: "LABEL"
+        description: "Reimbursement ID"
```
Comment (cuducos) [inline prometheus/postgres-exporter/businessInteligence.yml]: Not sure how this is going to figure in the UI, but this is useless from the UX perspective.

This ID is our Postgres-attributed unique identifier, and not the proper ID attributed by the Chamber of Deputies (which is not unique and is the `document_id`).

File: docker-compose.yml [hunk 19]
```diff
@@ -32,6 +32,56 @@ services:
   rosie:
     image: serenata/rosie
 
+  prometheus:
+    image: arthursens/serenata-prometheus
+    ports: 
+    - 9090:9090
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply docker-compose.yml]: No problem at all

File: docker-compose.yml [hunk 20]
```diff
@@ -32,6 +32,56 @@ services:
...
+    image: wrouesnel/postgres_exporter
+    environment:
+      DATA_SOURCE_NAME: postgres://jarbas:mysecretpassword@postgres/jarbas?sslmode=disable
+    links:
+    - postgres:postgres
+    volumes:
+    - ./prometheus/postgres-exporter:/prometheus
+    command: --disable-default-metrics --disable-settings-metrics --extend.query-path /prometheus/businessInteligence.yml
+
+
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply docker-compose.yml]: I didn't get what you mean by services D:

You mean changing
```
command: --disable-default-metrics --disable-settings-metrics --extend.query-path /prometheus/businessInteligence.yml
```
To
```
command: ["--disable-default-metrics", "--disable-settings-metrics", "--extend.query-path", "/prometheus/businessInteligence.yml"]
```

Or changing 
```
links:
    - node_exporter:node_exporter
    - process_exporter:process_exporter
    - postgres_exporter:postgres_exporter
    - bi_exporter:bi_exporter
```
To
```
links: ["node_exporter:node_exporter", "process_exporter:process_exporter", "postgres_exporter:postgres_exporter", "bi_exporter:bi_exporter" ]
```

[the remark this answers is not in the document]
Comment (ArthurSens) [reply docker-compose.yml]: Sure thing!

File: prometheus/postgres-exporter/businessInteligence.yml [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+    - year: 
+        usage: "LABEL"
+        description: "Year that the reimbursement was made"
+    - month:
+        usage: "LABEL"
+        description: "Month that the reimbursement was made"
+    - state:
+        usage: "LABEL"
+        description: "Federal state that the reimbursement was made"
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply prometheus/postgres-exporter/businessInteligence.yml]: Oh my bad

File: prometheus/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,5 @@
...
+
```
[the remark this answers is not in the document]
Comment (ArthurSens) [reply prometheus/Dockerfile]: No we do not, I'll erase it

File: prometheus/postgres-exporter/businessInteligence.yml (hunk 18 above)
Comment (ArthurSens) [reply prometheus/postgres-exporter/businessInteligence.yml]: It's actually kind useless, but not really

Similar to a relational database, prometheus can't store metrics with same name and labels(Relational dbs can't have duplicated primary keys).

All the registries queried by this query will have the same metric name (chamber_of_deputies_reimbursement_total_net_value), and the labels will vary according with their columns values....

I've noticed that if we don't query that ID column, we will have duplicated metrics with the exactly same labels, which will cause an error and we wouldn't be able to collect those metrics.

In short, we need to collect that ID just to avoid some errors, but its not used for anything

**Comment by ArthurSens** [comment]
> Over all I think it's a interesting contribution. Also I'm very glad for the effort you've put on the PR – not only in the code, but in the description and comments. Pretty rare to see outstanding documentation of the intended changes : )

Thanks! I just love projects and tools that have outstanding documentation like yours and I try to do the same.

> As far as I got, this PR implements a party analytics, that is something Serenata has never done for some very specific reasons:

> Nature of data: we (used to) belive in social control of politicians, and they change parties more than we would like they did, so focusing on parties is a focsung on somethng relatively volatile when compared to focusing on people
Representation bias: different parties have different number of representatives in the house, thus parties with more elected congresspeople will be more on the spotlight of any visualization or comparison that groups reimburements by parties (the same is valid for grouping by state, by the way)
Thus I would like to invite other maintainer to think and rethink how useful are these visualizations and what risks they imply in terms of interpretation, bias ans skewed meaning of data.

Yes, I've built a dashboard focused on parties but no, that's not the main point of this PR. I want to provide a place where you guys go to when you want to develop analytics UI and I've built one single dashboard just for demonstration(It took me 15m to build it). Grafana is soooo easy to learn, it's open source and also have excellent documentation, so anyone could create new dashboards and easly add it to grafana just by adding their JSON files at the provisioning folder.

Just in case you are new to grafana... We don't need to write the JSON. We build the dashboard using the UI and then we can export it to a JSON file.

image


> Pragmatic: complexity of the code base

> In spite of the great documentation of the PR, no file makes this documentation part of the code base. Current and future users would have to go back to this PR to understand part of the docker-compose.yml and the grafana/ and prometeus/ directories. And it's a lot of code (loooong JSONs and a bunch of YAMLs specifing how this dashboard should work).

> I do think a more experimental use os Prometheus & Grafana can be interesting for exporing this data, and I'm not asing for a use case, for example. However I think most of what is documented in this PR should feature, for example, in (so far, non-existng) grafana/README.md and prometeus/README.md. Also I think the services added to docker-compose.yml should feature somewhere in README.md or CONTRIBUTING.md, for example.

Hmmmm, how could I forget that? :thinking: 
I strongly agree that a `README.md` is necessary so more people feel confortable contrinuting with this. I will try to write them ASAP

File: prometheus/postgres-exporter/businessInteligence.yml (hunk 18 above)
Comment (cuducos) [reply prometheus/postgres-exporter/businessInteligence.yml]: Got it! Many thanks for the explanation. Maybe we could just rename the label so whoever uses it knows it's an id attributed by Serenata and not by the Chamber of Deputies. Maybe _Jarbas ID_ or a longer version _ID attributed to the reimbursement by in Serenata_…

**Comment by ArthurSens** [comment]
I do have some end of semester exams to study to and some others stuff behind schedule at work :worried: so I won't be able to advance with this PR right now....

I will try to save some time on the weekend for it

**Comment by cuducos** [comment]
That's awesome, @ArthurSens. I think my role a volunteer code reviewer is done now. I'll let @sergiomario and the Open Knowledge crew ponder on the questions I raised : )

Once more, many thanks for the awesome PR.

Mario, LGMT counting on the forthcoming changes Arthur mentioned in the conversation, and taking into account my disclaimer (i just reviewed code, haven't tested the new apps).

**Comment by ArthurSens** [comment]
Just commenting on this last commit...

I've made the changes you asked on the docker-compose regarding it's identation aaaaaand I've changed the images of prometheus and grafana to their originals...

I think that having to build a new image everytime a new dashboard is build will have a bad impact with contributions, so maybe using external volumes would be a better option here.

With this option you can change prometheus configuration just by editting prometheus.yml in your host machine... Adding new dashboards follows the same logic, add the JSON file to the monitoring folder, no need to make a new image after that


I will work on the README.md on the next few days...I hope I can finish it by the next weekend

**Comment by cuducos** [comment]
> I think that having to build a new image everytime a new dashboard is build will have a bad impact with contributions, so maybe using external volumes would be a better option here.

Awesome ❤️

**Comment by ArthurSens** [comment]
Documentation created...
I've tried to follow Jarbas and Rosie's README as example when structuring prometheus and grafana's README, but please tell me if there are changes that I still need to make
